File: src/storage/pg-vector.ts
import { randomUUID } from "node:crypto";
import type { PoolConfig } from "pg";
import type { JobEntry, LogEntry, MemoryEntry, ScoredMemory, StorageBackend } from "./types.js";

export class PgVectorBackend implements StorageBackend {
  private pool: import("pg").Pool | null = null;
  private hasVector = false;

  public constructor(private readonly config: PoolConfig) {}

  public async initialize(): Promise<void> {
    const { default: pg } = await import("pg");
    this.pool = new pg.Pool(this.config);

    const client = await this.pool.connect();
    try {
      try {
        await client.query("CREATE EXTENSION IF NOT EXISTS vector");
        this.hasVector = true;
      } catch {
        this.hasVector = false;
      }

      await client.query(`
        CREATE TABLE IF NOT EXISTS memories (
          id TEXT PRIMARY KEY,
          session_id TEXT NOT NULL,
          content TEXT NOT NULL,
          role TEXT NOT NULL,
          created_at TIMESTAMPTZ DEFAULT NOW(),
          embedding TEXT
        )
      `);

      if (this.hasVector) {
        await client.query(`
          ALTER TABLE memories ADD COLUMN IF NOT EXISTS embedding_vec vector(1536)
        `).catch(() => {});
        await client.query(`
          CREATE INDEX IF NOT EXISTS memories_vec_idx
            ON memories USING ivfflat (embedding_vec vector_cosine_ops)
        `).catch(() => {});
      }

      await client.query(`
        CREATE TABLE IF NOT EXISTS jobs (
          id TEXT PRIMARY KEY,
          session_id TEXT NOT NULL,
          status TEXT NOT NULL DEFAULT 'pending',
          type TEXT NOT NULL,
          payload JSONB,
          result JSONB,
          error TEXT,
          created_at TIMESTAMPTZ DEFAULT NOW(),
          updated_at TIMESTAMPTZ DEFAULT NOW()
        )
      `);

      await client.query(`
        CREATE TABLE IF NOT EXISTS audit_logs (
          id TEXT PRIMARY KEY,
          timestamp TIMESTAMPTZ NOT NULL,
          event TEXT NOT NULL,
          actor TEXT NOT NULL,
          session_id TEXT,
          data JSONB
        )
      `);
    } finally {
      client.release();
    }
  }

  private get db(): import("pg").Pool {
    if (!this.pool) throw new Error("PgVectorBackend not initialized");
    return this.pool;
  }

  public async saveMemory(entry: Omit<MemoryEntry, "id" | "createdAt">): Promise<MemoryEntry> {
    const id = randomUUID();
    const embeddingJson = entry.embedding ? JSON.stringify(entry.embedding) : null;
    await this.db.query(
      `INSERT INTO memories (id, session_id, content, role, embedding) VALUES ($1, $2, $3, $4, $5)`,
      [id, entry.sessionId, entry.content, entry.role, embeddingJson],
    );
    if (this.hasVector && entry.embedding) {
      await this.db.query(
        `UPDATE memories SET embedding_vec = $1::vector WHERE id = $2`,
        [`[${entry.embedding.join(",")}]`, id],
      ).catch(() => {});
    }
    return { ...entry, id, createdAt: new Date() };
  }

  public async searchByVector(embedding: number[], sessionId?: string, limit = 10): Promise<ScoredMemory[]> {
    if (!this.hasVector) return [];
    const vec = `[${embedding.join(",")}]`;
    const params: unknown[] = [vec, limit];
    const where = sessionId ? "AND session_id = $3" : "";
    if (sessionId) params.push(sessionId);
    const res = await this.db.query(
      `SELECT *, 1 - (embedding_vec <=> $1::vector) AS score
       FROM memories
       WHERE embedding_vec IS NOT NULL ${where}
       ORDER BY embedding_vec <=> $1::vector
       LIMIT $2`,
      params,
    );
    return res.rows.map((r) => this.rowToMemory(r, parseFloat(r.score), "vector"));
  }

  public async searchByFTS(query: string, sessionId?: string, limit = 10): Promise<ScoredMemory[]> {
    const params: unknown[] = [query, limit];
    const where = sessionId ? "AND session_id = $3" : "";
    if (sessionId) params.push(sessionId);
    const res = await this.db.query(
      `SELECT *, ts_rank(to_tsvector('english', content), plainto_tsquery('english', $1)) AS score
       FROM memories
       WHERE to_tsvector('english', content) @@ plainto_tsquery('english', $1) ${where}
       ORDER BY score DESC
       LIMIT $2`,
      params,
    );
    return res.rows.map((r) => this.rowToMemory(r, parseFloat(r.score), "fts"));
  }

  public async listMemories(sessionId: string, limit = 50): Promise<MemoryEntry[]> {
    const res = await this.db.query(
      `SELECT * FROM memories WHERE session_id = $1 ORDER BY created_at DESC LIMIT $2`,
      [sessionId, limit],
    );
    return res.rows.map((r) => this.rowToMemory(r, 0, "fts"));
  }

  public async saveJob(entry: Omit<JobEntry, "id" | "createdAt" | "updatedAt">): Promise<JobEntry> {
    const id = randomUUID();
    const now = new Date();
    await this.db.query(
      `INSERT INTO jobs (id, session_id, status, type, payload) VALUES ($1, $2, $3, $4, $5)`,
      [id, entry.sessionId, entry.status, entry.type, JSON.stringify(entry.payload ?? null)],
    );
    return { ...entry, id, createdAt: now, updatedAt: now };
  }

  public async updateJob(id: string, update: Partial<Pick<JobEntry, "status" | "result" | "error">>): Promise<void> {
    const sets: string[] = ["updated_at = NOW()"];
    const params: unknown[] = [];
    let i = 1;
    if (update.status !== undefined) { sets.push(`status = $${i++}`); params.push(update.status); }
    if (update.result !== undefined) { sets.push(`result = $${i++}`); params.push(JSON.stringify(update.result)); }
    if (update.error !== undefined) { sets.push(`error = $${i++}`); params.push(update.error); }
    params.push(id);
    await this.db.query(`UPDATE jobs SET ${sets.join(", ")} WHERE id = $${i}`, params);
  }

  public async listJobs(sessionId?: string): Promise<JobEntry[]> {
    const res = sessionId
      ? await this.db.query(`SELECT * FROM jobs WHERE session_id = $1 ORDER BY created_at DESC`, [sessionId])
      : await this.db.query(`SELECT * FROM jobs ORDER BY created_at DESC`);
    return res.rows.map((r) => ({
      id: r.id,
      sessionId: r.session_id,
      status: r.status,
      type: r.type,
      payload: r.payload,
      result: r.result,
      error: r.error,
      createdAt: new Date(r.created_at),
      updatedAt: new Date(r.updated_at),
    }));
  }

  public async saveLog(entry: Omit<LogEntry, "id">): Promise<void> {
    await this.db.query(
      `INSERT INTO audit_logs (id, timestamp, event, actor, session_id, data) VALUES ($1, $2, $3, $4, $5, $6)`,
      [randomUUID(), entry.timestamp, entry.event, entry.actor, entry.sessionId ?? null, JSON.stringify(entry.data ?? null)],
    );
  }

  public async listLogs(sessionId?: string, limit = 100): Promise<LogEntry[]> {
    const res = sessionId
      ? await this.db.query(`SELECT * FROM audit_logs WHERE session_id = $1 ORDER BY timestamp DESC LIMIT $2`, [sessionId, limit])
      : await this.db.query(`SELECT * FROM audit_logs ORDER BY timestamp DESC LIMIT $1`, [limit]);
    return res.rows.map((r) => ({
      id: r.id,
      timestamp: new Date(r.timestamp),
      event: r.event,
      actor: r.actor,
      sessionId: r.session_id ?? undefined,
      data: r.data,
    }));
  }

  private rowToMemory(r: Record<string, unknown>, score: number, source: "vector" | "fts"): ScoredMemory {
    return {
      id: r.id as string,
      sessionId: r.session_id as string,
      content: r.content as string,
      role: r.role as string,
      createdAt: new Date(r.created_at as string),
      embedding: r.embedding ? JSON.parse(r.embedding as string) : undefined,
      score,
      source,
    };
  }
}

File: titanx/storage/pg_vector.py
from __future__ import annotations

import json
from datetime import datetime, timezone
from typing import Any
from uuid import uuid4

from .types import JobEntry, LogEntry, MemoryEntry, ScoredMemory, StorageBackend


class PgVectorBackend(StorageBackend):
    def __init__(self, dsn: str) -> None:
        self._dsn = dsn
        self._pool = None
        self._has_vector = False

    async def initialize(self) -> None:
        import asyncpg
        self._pool = await asyncpg.create_pool(self._dsn)
        async with self._pool.acquire() as conn:
            try:
                await conn.execute("CREATE EXTENSION IF NOT EXISTS vector")
                self._has_vector = True
            except Exception:
                self._has_vector = False

            await conn.execute("""
                CREATE TABLE IF NOT EXISTS memories (
                    id TEXT PRIMARY KEY,
                    session_id TEXT NOT NULL,
                    content TEXT NOT NULL,
                    role TEXT NOT NULL,
                    created_at TIMESTAMPTZ DEFAULT NOW(),
                    embedding TEXT
                )
            """)

            if self._has_vector:
                try:
                    await conn.execute("ALTER TABLE memories ADD COLUMN IF NOT EXISTS embedding_vec vector(1536)")
                    await conn.execute("""
                        CREATE INDEX IF NOT EXISTS memories_vec_idx
                        ON memories USING ivfflat (embedding_vec vector_cosine_ops)
                    """)
                except Exception:
                    pass

            await conn.execute("""
                CREATE TABLE IF NOT EXISTS jobs (
                    id TEXT PRIMARY KEY,
                    session_id TEXT NOT NULL,
                    status TEXT NOT NULL DEFAULT 'pending',
                    type TEXT NOT NULL,
                    payload JSONB,
                    result JSONB,
                    error TEXT,
                    created_at TIMESTAMPTZ DEFAULT NOW(),
                    updated_at TIMESTAMPTZ DEFAULT NOW()
                )
            """)

            await conn.execute("""
                CREATE TABLE IF NOT EXISTS audit_logs (
                    id TEXT PRIMARY KEY,
                    timestamp TIMESTAMPTZ NOT NULL,
                    event TEXT NOT NULL,
                    actor TEXT NOT NULL,
                    session_id TEXT,
                    data JSONB
                )
            """)

    async def save_memory(self, session_id: str, content: str, role: str, embedding: list[float] | None = None) -> MemoryEntry:
        assert self._pool
        id_ = str(uuid4())
        embedding_json = json.dumps(embedding) if embedding else None
        async with self._pool.acquire() as conn:
            await conn.execute(
                "INSERT INTO memories (id, session_id, content, role, embedding) VALUES ($1, $2, $3, $4, $5)",
                id_, session_id, content, role, embedding_json,
            )
            if self._has_vector and embedding:
                vec_str = f"[{','.join(str(x) for x in embedding)}]"
                try:
                    await conn.execute(
                        "UPDATE memories SET embedding_vec = $1::vector WHERE id = $2",
                        vec_str, id_,
                    )
                except Exception:
                    pass
        return MemoryEntry(id=id_, session_id=session_id, content=content, role=role,
                           created_at=datetime.now(timezone.utc), embedding=embedding)

    async def search_by_vector(self, embedding: list[float], session_id: str | None = None, limit: int = 10) -> list[ScoredMemory]:
        if not self._has_vector:
            return []
        assert self._pool
        vec_str = f"[{','.join(str(x) for x in embedding)}]"
        where = "AND session_id = $3" if session_id else ""
        params: list[Any] = [vec_str, limit]
        if session_id:
            params.append(session_id)
        async with self._pool.acquire() as conn:
            rows = await conn.fetch(
                f"""SELECT *, 1 - (embedding_vec <=> $1::vector) AS score
                    FROM memories
                    WHERE embedding_vec IS NOT NULL {where}
                    ORDER BY embedding_vec <=> $1::vector
                    LIMIT $2""",
                *params,
            )
        return [self._row_to_scored(r, float(r["score"]), "vector") for r in rows]

    async def search_by_fts(self, query: str, session_id: str | None = None, limit: int = 10) -> list[ScoredMemory]:
        assert self._pool
        where = "AND session_id = $3" if session_id else ""
        params: list[Any] = [query, limit]
        if session_id:
            params.append(session_id)
        async with self._pool.acquire() as conn:
            rows = await conn.fetch(
                f"""SELECT *,
                       ts_rank(to_tsvector('english', content), plainto_tsquery('english', $1)) AS score
                    FROM memories
                    WHERE to_tsvector('english', content) @@ plainto_tsquery('english', $1) {where}
                    ORDER BY score DESC
                    LIMIT $2""",
                *params,
            )
        return [self._row_to_scored(r, float(r["score"]), "fts") for r in rows]

    async def list_memories(self, session_id: str, limit: int = 50) -> list[MemoryEntry]:
        assert self._pool
        async with self._pool.acquire() as conn:
            rows = await conn.fetch(
                "SELECT * FROM memories WHERE session_id = $1 ORDER BY created_at DESC LIMIT $2",
                session_id, limit,
            )
        return [self._row_to_memory(r) for r in rows]

    async def save_job(self, session_id: str, type: str, status: str = "pending", payload: Any = None) -> JobEntry:
        assert self._pool
        id_ = str(uuid4())
        now = datetime.now(timezone.utc)
        async with self._pool.acquire() as conn:
            await conn.execute(
                "INSERT INTO jobs (id, session_id, status, type, payload) VALUES ($1, $2, $3, $4, $5)",
                id_, session_id, status, type, json.dumps(payload) if payload is not None else None,
            )
        return JobEntry(id=id_, session_id=session_id, status=status, type=type,
                        created_at=now, updated_at=now, payload=payload)

    async def update_job(self, id: str, status: str | None = None, result: Any = None, error: str | None = None) -> None:
        assert self._pool
        sets = ["updated_at = NOW()"]
        params: list[Any] = []
        i = 1
        if status is not None:
            sets.append(f"status = ${i}"); params.append(status); i += 1
        if result is not None:
            sets.append(f"result = ${i}"); params.append(json.dumps(result)); i += 1
        if error is not None:
            sets.append(f"error = ${i}"); params.append(error); i += 1
        params.append(id)
        async with self._pool.acquire() as conn:
            await conn.execute(f"UPDATE jobs SET {', '.join(sets)} WHERE id = ${i}", *params)

    async def list_jobs(self, session_id: str | None = None) -> list[JobEntry]:
        assert self._pool
        async with self._pool.acquire() as conn:
            if session_id:
                rows = await conn.fetch("SELECT * FROM jobs WHERE session_id = $1 ORDER BY created_at DESC", session_id)
            else:
                rows = await conn.fetch("SELECT * FROM jobs ORDER BY created_at DESC")
        return [self._row_to_job(r) for r in rows]

    async def save_log(self, timestamp: datetime, event: str, actor: str, session_id: str | None = None, data: Any = None) -> None:
        assert self._pool
        async with self._pool.acquire() as conn:
            await conn.execute(
                "INSERT INTO audit_logs (id, timestamp, event, actor, session_id, data) VALUES ($1, $2, $3, $4, $5, $6)",
                str(uuid4()), timestamp, event, actor, session_id, json.dumps(data) if data is not None else None,
            )

    async def list_logs(self, session_id: str | None = None, limit: int = 100) -> list[LogEntry]:
        assert self._pool
        async with self._pool.acquire() as conn:
            if session_id:
                rows = await conn.fetch(
                    "SELECT * FROM audit_logs WHERE session_id = $1 ORDER BY timestamp DESC LIMIT $2",
                    session_id, limit,
                )
            else:
                rows = await conn.fetch("SELECT * FROM audit_logs ORDER BY timestamp DESC LIMIT $1", limit)
        return [LogEntry(id=r["id"], timestamp=r["timestamp"], event=r["event"],
                         actor=r["actor"], session_id=r["session_id"], data=r["data"]) for r in rows]

    def _row_to_scored(self, r: Any, score: float, source: str) -> ScoredMemory:
        m = self._row_to_memory(r)
        return ScoredMemory(**m.__dict__, score=score, source=source)

    def _row_to_memory(self, r: Any) -> MemoryEntry:
        emb = json.loads(r["embedding"]) if r["embedding"] else None
        return MemoryEntry(
            id=r["id"], session_id=r["session_id"], content=r["content"],
            role=r["role"], created_at=r["created_at"], embedding=emb,
        )

    def _row_to_job(self, r: Any) -> JobEntry:
        return JobEntry(
            id=r["id"], session_id=r["session_id"], status=r["status"],
            type=r["type"], payload=r["payload"], result=r["result"],
            error=r["error"], created_at=r["created_at"], updated_at=r["updated_at"],
        )
